proxy/src/telemetry/metrics/labels/mod.rs

```diff
@@ -441,6 +491,68 @@ impl Into<ctx::transport::TlsStatus> for TlsStatus {
     }
 }
 
+impl HandshakeFailLabels {
+    pub fn proxy(ctx: &ctx::transport::Ctx,
+                 err: &connection::HandshakeError)
+               -> Self {
+        HandshakeFailLabels::Proxy {
+            transport: TransportLabels::new(ctx),
+            reason: err.into(),
+        }
+    }
+
+    pub fn control(ctx: &event::ControlConnection,
+                   err: &connection::HandshakeError)
+                   -> Self {
+        let (peer, remote_addr) = match ctx {
+            event::ControlConnection::Accept { remote_addr, .. } =>
+                (Peer::Dst, *remote_addr),
+            event::ControlConnection::Connect { remote_addr } =>
+                (Peer::Src, *remote_addr),
+        };
+        HandshakeFailLabels::Control {
+            peer,
+            remote_addr,
+            reason: err.into(),
+        }
+    }
+}
+
+impl fmt::Display for HandshakeFailLabels {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            HandshakeFailLabels::Proxy { transport, reason } =>
+                write!(f, "{},{}", transport, reason),
+            HandshakeFailLabels::Control { peer, remote_addr, reason } =>
+                write!(f, "{},remote=\"{}\",{}",peer, remote_addr, reason),
+        }
+    }
+}
+
+impl fmt::Display for HandshakeFailReason {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            HandshakeFailReason::Tls(ref reason) =>
+                write!(f, "reason=\"tls_error\",tls_error=\"{}\"", reason),
+            HandshakeFailReason::Io(Some(ref errno)) =>
+                write!(f, "reason=\"io_error\",errno=\"{}\"", errno),
+            HandshakeFailReason::Io(None) =>
+                f.pad("reason=\"io_error\",errno=\"unknown\""),
+        }
+    }
+}
+
+impl<'a> From<&'a connection::HandshakeError> for HandshakeFailReason {
```

Comment:
It seems likely that when we move `HandshakeError` to this module then we can combine it with `HandshakeFailReason`. Notice in particular that `HandshakeError::from` splits a single `io::Error` into two variants and then this function deconstructs those two variants.